From ea7483abeb3307bfc478c247da8817cb9396d74e Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:05:07 -0400
Subject: [ethernet] qlcnic: Fix SR-IOV configuration

Message-id: <1408626422-3941-24-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88551
O-Subject: [PATCH RHEL 7.1 023/138] qlcnic: Fix SR-IOV configuration
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Manish Chopra <manish.chopra@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 1ed98ed55d6bf47d5a21b1e2db35ceb8b9a4c91c

o Interface needs to be brought down and up while configuring SR-IOV.
  Protect interface up/down using rtnl_lock()/rtnl_unlock()

Signed-off-by: Manish Chopra <manish.chopra@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
index 330d9a8..686f460 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
@@ -397,6 +397,7 @@ static int qlcnic_pci_sriov_disable(struct qlcnic_adapter *adapter)
 {
  struct net_device *netdev = adapter->netdev;
 
+ rtnl_lock();
  if (netif_running(netdev))
   __qlcnic_down(adapter, netdev);
 
@@ -407,12 +408,15 @@ static int qlcnic_pci_sriov_disable(struct qlcnic_adapter *adapter)
  /* After disabling SRIOV re-init the driver in default mode
     configure opmode based on op_mode of function
   */
- if (qlcnic_83xx_configure_opmode(adapter))
+ if (qlcnic_83xx_configure_opmode(adapter)) {
+  rtnl_unlock();
   return -EIO;
+ }
 
  if (netif_running(netdev))
   __qlcnic_up(adapter, netdev);
 
+ rtnl_unlock();
  return 0;
 }
 
@@ -533,6 +537,7 @@ static int qlcnic_pci_sriov_enable(struct qlcnic_adapter *adapter, int num_vfs)
   return -EIO;
  }
 
+ rtnl_lock();
  if (netif_running(netdev))
   __qlcnic_down(adapter, netdev);
 
@@ -555,6 +560,7 @@ static int qlcnic_pci_sriov_enable(struct qlcnic_adapter *adapter, int num_vfs)
   __qlcnic_up(adapter, netdev);
 
 error:
+ rtnl_unlock();
  return err;
 }
 
-- 
1.7.1